keps/sig-api-machinery/3617-fine-grained-authz/README.md

```diff
@@ -0,0 +1,1162 @@
+	// (scope and select) objects. May match selectors of replication controllers
+	// and services.
+	// More info: http://kubernetes.io/docs/user-guide/labels
+	// +optional
+	Labels map[string]string `json:"labels,omitempty" protobuf:"bytes,11,rep,name=labels"`
+	...
+```
+
+This is very inconvenient as it leaves only one place to declare things in the
+schema. Note that we must not make any incompatible changes to the serialization
+of this type. This is one way to make this work:
+
+```go
+	...
+	Labels map[LabelKey]LabelValue `json:"labels,omitempty" protobuf:"bytes,11,rep,name=labels"`
+	...
+}
+.
+.
+.
+
+// LabelKey is a key for a label; it is a string.
+// +permission-parameter-source:"label-key"
+type LabelKey string
+
+// LabelValue is a value in a label; it is a string.
+// +permission-verb:"label"
+// +permission-parameterized-by:"label-key"
+// +permission-parameter-treatment:slash-delimited-prefix
+type LabelValue string
+```
+
+This says to the permission system that in order to change a LabelValue, there
+must be something in the field path emitting a "label-key" parameter; that that
+value needs to be massaged by a "slash-delimited-prefix" function; and that the
+permission should then be checked with that parameter.
+
+Supposing the key is "mycompany.example.com/FooLabel", The authz system will see
+the verb "granular:label(mycompany.example.com)".
+
+DECISION NEEDED: would it be better to use a CEL expression to process the
+parameter?
+
+#### Specific Finalizers
+
+TODO (similar choice).
+
+#### Specific Conditions
+
+TODO (similar choice).
+
+#### Complete List of parameter treatments
+
+* None (omit the tag completely). The referenced field is used verbatim as the parameter.
+* `slash-delimited-prefix`. The referenced field is split by '/' characters and
+  the first segment is used as the parameter.
+* (in progress)
+
+#### Style Guide / Consistency Help
+
+Rules for making up a permission:
+
+* The permission name SHOULD be a noun.
+* The permission name MUST be in lowerCamelCase.
+* The permission name SHOULD refer to a conceptual attribute, not a field name.
+  This helps the permission be reusable in other contexts with different
+  enclosing parent fields.
+* The permission MUST NOT reference ancestor fields ("nodeAssigment" not
+  "pod.nodeAssignment")
+
+We will update the documentation generation to list the permissions both with
+the fields / types they guard, as well as in a standalone list for easy
+browsing. This list will help API reviewers keep the permissions on built-in
+resources coherent.
+
+DECISION NEEDED:
+
+Option A: we will put a registry file in the kubernetes repository so that 3rd
+party extension authors can register their custom permissions and dedup/reuse
+rather than create new ones, if possible.
+
+Option B: we will require 3rd party authors to prefix their custom permissions
+with the string "ext" (TODO: "mycompany.example.com" if there is likely to be
+space).
```

Comment:
```suggestion
<<[UNRESOLVED decision needed]>>
Option A: we will put a registry file in the k/kubernetes repository so that 3rd
party extension authors can register their custom permissions and dedup/reuse
rather than create new ones, if possible.

Option B: we will require 3rd party authors to prefix their custom permissions
with the string "ext" (TODO: "mycompany.example.com" if there is likely to be
space).
<<[/UNRESOLVED]>>
```

My option C would be to copy the idea from labels and use `kubernetes.io/whatever` for permissions defined by Kubernetes. Prefixes need to be valid DNS subdomains and should be associated with the party that defines them. Etc.